src/main/java/dev/aziz/grocerystore/services/BasketItemService.java

```diff
@@ -0,0 +1,99 @@
+package dev.aziz.grocerystore.services;
+
+import dev.aziz.grocerystore.dtos.BasketItemDto;
+import dev.aziz.grocerystore.dtos.BasketTotalDto;
+import dev.aziz.grocerystore.entities.BasketItem;
+import dev.aziz.grocerystore.entities.Item;
+import dev.aziz.grocerystore.entities.User;
+import dev.aziz.grocerystore.exceptions.AppException;
+import dev.aziz.grocerystore.mappers.BasketItemMapper;
+import dev.aziz.grocerystore.repositories.BasketItemRepository;
+import dev.aziz.grocerystore.repositories.ItemRepository;
+import dev.aziz.grocerystore.repositories.UserRepository;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor
+@Service
+@Slf4j
+public class BasketItemService {
+
+    private final BasketItemRepository basketItemRepository;
+    private final ItemRepository itemRepository;
+    private final UserRepository userRepository;
+    private final BasketItemMapper basketItemMapper;
+
+
+    public BasketItem addItem(String login, Long itemId, Integer amount) {
+        Item item = itemRepository.findItemById(itemId)
+                .orElseThrow(() -> new AppException("Item not found.", HttpStatus.NOT_FOUND));
+
+        User user = userRepository.findByLogin(login)
+                .orElseThrow(() -> new AppException("User not found.", HttpStatus.NOT_FOUND));
+        BasketItem saved = basketItemRepository.save(
+                BasketItem.builder()
+                        .user(user)
+                        .item(item)
+                        .amount(amount)
+                        .build()
+        );
+        return saved;
+    }
+
+    public BasketTotalDto getItems(String login) {
+        log.info("User {} is retrieving items from the basket.", login);
+
+        User user = userRepository.findByLogin(login)
+                .orElseThrow(() -> new AppException("User not found.", HttpStatus.NOT_FOUND));
+        List<BasketItem> basketByUser = basketItemRepository.findBasketsByUserId(user.getId())
+                .orElseThrow(() -> new AppException("Basket not found.", HttpStatus.NOT_FOUND));
+
+        List<BasketItemDto> basketItemDtoList = new ArrayList<>();
+        for (BasketItem basketItem : basketByUser) {
+            basketItemDtoList.add(basketItemMapper.basketItemToBasketItemDto(basketItem));
+        }
+
+        Map<String, List<BasketItemDto>> groupedItems = basketItemDtoList.stream()
```

Comment:
To avoid this complexity, in the addItem method, sum the values to avoid duplicated items.